src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -0,0 +1,513 @@
+module dyn_mpas_subdriver
+    !-------------------------------------------------------------------------------
+    ! module dyn_mpas_subdriver
+    !
+    ! This module manages the life cycle (i.e., initialization, running, and
+    ! finalization) of MPAS as a dynamical core within CAM-SIMA.
+    !
+    !-------------------------------------------------------------------------------
+
+    use, intrinsic :: iso_fortran_env, only: output_unit
+
+    ! Modules from external libraries.
+    use mpi, only: mpi_comm_null, mpi_comm_rank, mpi_success
+    use pio, only: iosystem_desc_t, pio_iosystem_is_active
+
+    ! Modules from MPAS.
+    use atm_core_interface, only: atm_setup_core, atm_setup_domain
+    use mpas_derived_types, only: core_type, domain_type
+    use mpas_domain_routines, only: mpas_allocate_domain
+    use mpas_framework, only: mpas_framework_init_phase1, mpas_framework_init_phase2
+    use mpas_kind_types, only: strkind
+    use mpas_pool_routines, only: mpas_pool_get_config
+
+    implicit none
+
+    private
+    public :: mpas_dynamical_core_type, mpas_dynamical_core
+
+    abstract interface
+        ! This interface is compatible with `endrun` from CAM-SIMA.
+        subroutine model_error_if(message, file, line)
+            character(*),           intent(in) :: message
+            character(*), optional, intent(in) :: file
+            integer,      optional, intent(in) :: line
+        end subroutine model_error_if
+    end interface
+
+    !> The "class" of MPAS dynamical core.
+    !> Important data structures like states of MPAS dynamical core are encapsulated inside this derived type to prevent misuse.
+    !> Type-bound procedures provide well-defined APIs for CAM-SIMA to interact with MPAS dynamical core.
+    type :: mpas_dynamical_core_type
+        private
+
+        integer :: log_unit = output_unit
+        integer :: mpi_comm = mpi_comm_null
+        integer :: mpi_rank = 0
+        logical :: mpi_rank_master = .false.
+
+        ! Actual implementation is supplied at runtime.
+        procedure(model_error_if), nopass, pointer :: model_error => null()
+
+        type(core_type), pointer :: corelist => null()
+        type(domain_type), pointer :: domain_ptr => null()
+    contains
+        private
+
+        procedure, pass, public :: debug_print => dyn_mpas_debug_print
+        procedure, pass, public :: init_phase1 => dyn_mpas_init_phase1
+        procedure, pass, public :: read_namelist => dyn_mpas_read_namelist
+        procedure, pass, public :: init_phase2 => dyn_mpas_init_phase2
+    end type mpas_dynamical_core_type
+
+    !> The "instance/object" of MPAS dynamical core.
+    type(mpas_dynamical_core_type) :: mpas_dynamical_core
+contains
+    !> Print a debug message with optionally the value(s) of a variable.
+    !> (KCW, 2024-02-03)
+    subroutine dyn_mpas_debug_print(self, message, variable)
```

Comment:
Is there a mechanism by which this routine can be effectively turned into a no-op so that debugging print statements aren't written to the log file except in cases where we're actually trying to debug the code?

What I'm wondering, more specifically, is whether CAM-SIMA provides a global debug flag that could be used here to activate the code here only when requested by that debug flag (which may be either a compile-time or a run-time option).

Reply:
CAM-SIMA uses CIME, so I think the standard mechanism to control debug/release builds is by toggling the `DEBUG` variable in `env_build.xml`. It causes either `DEBUG` macro (for debug builds) or `NDEBUG` macro (for release builds) to be defined during compilation [[1]](https://github.com/ESMCI/cime/blob/cime6.0.218/CIME/Tools/Makefile#L92).

I have guarded the subroutine body of `dyn_mpas_debug_print` inside a preprocessor directive. Debug prints are now disabled for release builds. See commit https://github.com/ESCOMP/CAM-SIMA/pull/248/commits/ce9bc5fbc3b224fc3ab552f02a68b3b7ea6a8f24.